# HG changeset patch
# User Punarbasu Purkayastha <ppurka@gmail.com>
# Date 1342359642 -28800
# Node ID bb9199b13a7f15fc205ae3f3b1d4466583667cc2
# Parent  69194615bd63053dcd9fee984f5a3e402c916656
 1. automatically exclude invalid points from a plot
 2. fix excluded regions in case more than one exclude point was in an invalid region

diff --git a/sage/plot/plot.py b/sage/plot/plot.py
--- a/sage/plot/plot.py
+++ b/sage/plot/plot.py
@@ -1266,6 +1266,25 @@
     else:
         data = generate_plot_points(f, xrange, plot_points, adaptive_tolerance, adaptive_recursion, randomize)
 
+    # Need exclude to be a list to be able to do automatic exclusion of
+    # plot regions.
+    if exclude is None:
+        exclude = []
+
+    for i in range(len(data)-1):
+        # If the difference between consecutive x-values is more than
+        # 3 times the difference between two consecutive plot points, then
+        # add an exclusion point. We assume that the data points are in
+        # increasing order of x-values.
+        if data[i+1][0] - data[i][0] > 3*(xmax - xmin)/plot_points:
+            exclude.append((data[i][0] + data[i+1][0])/2)
+
+    # We set exclude back to None if there are no points to be excluded
+    if exclude == []:
+        exclude = None
+    else:
+        exclude = sorted(exclude)
+
     if parametric:
         # We need the original x-values to be able to exclude points in parametric plots
         exclude_data = data
@@ -1375,11 +1394,13 @@
             if exclude is not None and (x0 <= exclusion_point <= x1):
                 G += line(data[start_index:i], **options)
                 start_index = i + 2
-                try:
-                    exclusion_point = exclude.pop()
-                except IndexError:
-                    # all excluded points were considered
-                    exclude = None
+                while exclusion_point <= x1:
+                    try:
+                        exclusion_point = exclude.pop()
+                    except IndexError:
+                        # all excluded points were considered
+                        exclude = None
+                        break
 
         G += line(data[start_index:], legend_label=legend_label, **options)
     else:
